Version 2.0

ChristianRiesen opened the pull request:
Trying to get this to a better standard, requiring php 7.4

reedy commented:
I note for "us" (MediaWiki/Wikimedia), 7.2 would be appreciated

reedy commented on tests/bootstrap.php:
```diff
@@ -1,5 +1,5 @@
 <?php
```
Is this file actually needed for anything? It... shouldn't be...

reedy commented on tests/Base32Test.php:
```diff
@@ -1,51 +1,60 @@
 use Base32\Base32;
+use PHPUnit\Framework\TestCase;
+
+class Base32Test extends TestCase
```
Probably worth adding some @covers

reedy commented on composer.json:
```diff
@@ -14,12 +14,12 @@
     ],
     "require": {
-        "php": ">=5.3"
+        "php": "^7.4"
```
Most tend to do >=version atm...

So `>=7.4` would be fine (though, I would prefer `>=7.2`)

Or `^7.4||^8.0` would be better (or well, `^7.2||^8.0`)

reedy commented on composer.json:
```diff
@@ -14,12 +14,12 @@
-        "satooshi/php-coveralls": "^1.0"
+        "friendsofphp/php-cs-fixer": "^2.17",
+        "phpstan/phpstan": "^0.12",
+        "phpunit/phpunit": "^9.5"
```
`^8.5.13||^9.5.0` should work fine to keep 7.2 support, without additiona changes...

ChristianRiesen commented:
@reedy Thanks for your input. I wasn't aware MediaWiki is using this :)
I incorporated all your inputs and got the builds to work with all the versions (7.2 tests with old phpunit, everything else with the newer one). There was no functional need to make it require 7.4 and so I made sure it keeps working and being tested on 7.2 as well. Also getting tested on 8.0 so a move to that should be fine as well.

If you have more inputs, please send them my way, I'll leave this PR open for another day or two before merging and tagging. Not even a BC break happening here, so not a v2 after all :)

reedy commented:
> Not even a BC break happening here, so not a v2 after all :)

Well... YMMV on whether bumping PHP versions is a semver major change. Different groups argue different ways.

I would say as you're dropping 5.x, it's probably easier/better to just bump the major version; doing so is cheap. Rather than people having something like `^1.3`, still being on PHP 5.3 and wondering why 1.4 won't immediately work (though, composer should take of care of that, I think) :)

I'll have another look/review first

reedy reviewed (changes requested):
My only irk is the yoda conditionals - `'' == $foo` rather than `$foo == ''`, and more so the changing to them.. I understand this is personal style, and some projects (at least [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#yoda-conditions)) do actively suggest doing it.

Blockers: Couple of wording changes needed. And then consistency on whether `\` is used before global functions. Otherwise LGTM I think :)

ChristianRiesen commented on src/Base32.php:
```diff
@@ -136,11 +122,30 @@ public static function decode($base32String)
+    {
+        $binaryString = chunk_split($binaryString, $bits, ' ');
+
+        if (' ' === substr($binaryString, \strlen($binaryString) - 1)) {
```
It adds some speed, so why not use it consistently for everything. My IDE only flags certain functions where it is proven to add more speed, but might as well add it everywhere.

ChristianRiesen commented on README.md:
```diff
@@ -45,9 +37,9 @@ Have a RFC compliant Base32 encoder and decoder. The implementation could be imp
+Works on PHP 7.2 and later, including PHP 8.0.
 
-If you want to run the tests, PHPUnit 5.0+ or up is required. Tests require PHP 5.6 or 7.0+.
+Tests run on PHPUnit 9.5, with PHP 7.3 and later. For PHP 7.2 uses an older PHPUnit version.
```
Changed

ChristianRiesen commented:
Thanks again for the inputs @reedy 
I agree, consistency makes more sense, so since it can add performance, might as well do it. Also added a check to cs fixer for that for the future.
Also changed the README. Should be slightly better.
Regarding semver, I was of the opinion to make a major version change when changing minimum php version. After a discussion with a co worker I changed my mind. Composer takes care of that, as that is a dependency. The code itself (or it's interfaces really) should dictate a major change or not in the version. Thats at least my thinking here.
On yoda, I was not a fan at first, though saw the logical reasoining behind it. I have since starting to use it more seriously seen multiple bugs that could have been prevented by using it, so I'm convinced. It makes the conditions slightly harder to read, but this also points out to me when a condition is too large to properly understand quickly. Should have no influence of the functionality of the library though :)

I'll merge and tag then. More inputs always welcome. Now I need to fix my otp library :)
